### PR TITLE
core/device: ignore ID_PROCESSING udev property on enumerate

### LINKED ISSUE #35329: systemd daemon-reexec brings down device units if processing by udev (ID_PROCESSING=1)
### systemd version the issue has been seen with

256

### Used distribution

Centos 9

### Linux kernel version used

6.4

### CPU architectures issue was seen on

x86_64

### Component

systemd, systemd-udevd

## Background

At Meta, we have seen the following while upgrading systemd 255 -> 256 on a live Centos 9 system during our host provisioning process:
1. We have a service chef-firstboot-os.service that runs Chef and does the RPM upgrade.
3. Right after daemon-reexec and udevd restart (triggered by 255 -> 256 RPM specs), we see Chef receives a SIGTERM from systemd
4. Chef then is SIGKILLed 30s later due to TimeoutStopSec=30s for the unit

This occurs about 5-10% of the time and unfortunately I don't have a reliable local repro after a month - I can only reproduce on the host provisioning systems. The best I have is logs from PID1/udevd/Chef/bpftrace which I'm happy to share.

## Details

While upgrading systemd RPMs from 255 -> 256 under a single DNF transaction, we see Chef gets sent SIGTERM:
```
[2024-11-22T22:08:41.126-08:00] DEBUG: fb_dnf_package[multipackage for firstboot package updates] systemd not installed, installing 0:256.7-1.6.hs+fb.el9.x86_64
[2024-11-22T22:08:41.126-08:00] DEBUG: fb_dnf_package[multipackage for firstboot package updates] systemd-libs not installed, installing 0:256.7-1.6.hs+fb.el9.x86_64
[2024-11-22T22:08:41.126-08:00] DEBUG: fb_dnf_package[multipackage for firstboot package updates] systemd-container not installed, installing 0:256.7-1.6.hs+fb.el9.x86_64
[2024-11-22T22:08:41.126-08:00] DEBUG: fb_dnf_package[multipackage for firstboot package updates] systemd-pam not installed, installing 0:256.7-1.6.hs+fb.el9.x86_64
[2024-11-22T22:08:41.126-08:00] DEBUG: fb_dnf_package[multipackage for firstboot package updates] systemd-udev not installed, installing 0:256.7-1.6.hs+fb.el9.x86_64
[2024-11-22T22:08:41.127-08:00] DEBUG: fb_dnf_package[multipackage for firstboot package updates] systemd-networkd not installed, installing 0:256.7-1.6.hs+fb.el9.x86_64
[2024-11-22T22:10:36.164-08:00] DEBUG: SIGTERM received during converge, finishing converge to exit normally (send SIGINT to terminate immediately)
```

bpftrace confirms that systemd sends the SIGTERM but there is no `systemctl stop/kill` commands issued or dbus APIs called (monitored via bpftrace and busctl respectively):
```
["2024-11-23T14:27:13"] kill_pid_info: Process '/usr/lib/systemd/systemd\x00--system\x00--deserialize=110\x00' with PID '1' sending 15 to PID '13721'

        7f0506b0731d syscall+29 (/usr/lib64/libc.so.6)
        fffffff700000000 0xfffffff700000000 ([unknown])


        bpf_prog_df1df1d87cf4d0da_kfunc_vmlinux_kill_pid_info_6+1869
        bpf_prog_df1df1d87cf4d0da_kfunc_vmlinux_kill_pid_info_6+1869
        bpf_trampoline_6442543875+75
        kill_pid_info+5
        __se_sys_pidfd_send_signal+271
        do_syscall_64+61
        entry_SYSCALL_64_after_hwframe+70
```

Indeed systemd cancels all the jobs. But right before that we see systemd calling the catchup() handlers, done during reexec/reload, where many devices used by mounts are transitioned to the DEAD state!
```
Nov 22 22:10:35 twshared8602.05.cln3.facebook.com systemd[1]: Invoking unit catchup() handlers…
Nov 22 22:10:35 twshared8602.05.cln3.facebook.com systemd[1]: sys-devices-platform-serial8250-tty-ttyS11.device: Changed plugged -> dead
Nov 22 22:10:35 twshared8602.05.cln3.facebook.com systemd[1]: sys-devices-platform-serial8250-tty-ttyS4.device: Changed plugged -> dead
Nov 22 22:10:35 twshared8602.05.cln3.facebook.com systemd[1]: dev-disk-by\x2did-nvme\x2dKXG6AZNV512G_TOSHIBA_515F72OTFJL3\x2dpart2.device: Changed plugged -> dead
Nov 22 22:10:35 twshared8602.05.cln3.facebook.com systemd[1]: dev-disk-by\x2dlabel-ESP.device: Changed plugged -> dead
Nov 22 22:10:35 twshared8602.05.cln3.facebook.com systemd[1]: dev-disk-by\x2ddiskseq-17\x2dpart2.device: Changed plugged -> dead
Nov 22 22:10:35 twshared8602.05.cln3.facebook.com systemd[1]: dev-ttyS9.device: Changed plugged -> dead
Nov 22 22:10:35 twshared8602.05.cln3.facebook.com systemd[1]: sys-devices-platform-serial8250-tty-ttyS7.device: Changed plugged -> dead
Nov 22 22:10:35 twshared8602.05.cln3.facebook.com systemd[1]: dev-disk-by\x2dpath-pci\x2d0000:66:00.0\x2dnvme\x2d1\x2dpart2.device: Changed plugged -> dead
Nov 22 22:10:35 twshared8602.05.cln3.facebook.com systemd[1]: dev-disk-by\x2did-nvme\x2dKXG6AZNV512G_TOSHIBA_515F72OTFJL3\x2dpart1.device: Changed plugged -> dead
Nov 22 22:10:35 twshared8602.05.cln3.facebook.com systemd[1]: dev-ttyS8.device: Changed plugged -> dead
Nov 22 22:10:35 twshared8602.05.cln3.facebook.com systemd[1]: dev-ttyS10.device: Changed plugged -> dead
Nov 22 22:10:35 twshared8602.05.cln3.facebook.com systemd[1]: dev-disk-by\x2did-nvme\x2dKXG6AZNV512G_TOSHIBA_515F72OTFJL3_1\x2dpart2.device: Changed plugged -> dead
Nov 22 22:10:35 twshared8602.05.cln3.facebook.com systemd[1]: dev-nvme0n1p2.device: Changed plugged -> dead
...
Nov 22 22:10:35 twshared8602.05.cln3.facebook.com systemd[1]: workload.target: Job 9 workload.target/start finished, result=canceled
Nov 22 22:10:35 twshared8602.05.cln3.facebook.com systemd[1]: chef-firstboot.target: Job 29 chef-firstboot.target/start finished, result=canceled
Nov 22 22:10:35 twshared8602.05.cln3.facebook.com systemd[1]: metalos-report-workload.service: Job 10 metalos-report-workload.service/start finished, result=canceled
Nov 22 22:10:35 twshared8602.05.cln3.facebook.com systemd[1]: core-services.target: Job 28 core-services.target/start finished, result=canceled
Nov 22 22:10:35 twshared8602.05.cln3.facebook.com systemd[1]: workload-pre.target: Job 13 workload-pre.target/start finished, result=canceled
Nov 22 22:10:35 twshared8602.05.cln3.facebook.com systemd[1]: chef-firstboot-os.service: Job 224 chef-firstboot-os.service/start finished, result=canceled
Nov 22 22:10:35 twshared8602.05.cln3.facebook.com systemd[1]: chef-firstboot-tier.service: Job 226 chef-firstboot-tier.service/start finished, result=canceled
Nov 22 22:10:42 twshared8602.05.cln3.facebook.com systemd[1]: run-fs-control.mount: Job 5149 run-fs-control.mount/stop finished, result=canceled
Nov 22 22:10:42 twshared8602.05.cln3.facebook.com systemd[1]: systemd-tmpfiles-setup.service: Job 5150 systemd-tmpfiles-setup.service/stop finished, result=canceled
Nov 22 22:10:42 twshared8602.05.cln3.facebook.com systemd[1]: var-facebook-tupperware.mount: Job 5130 var-facebook-tupperware.mount/stop finished, result=canceled
```

We can see at this time, udev failed and then restarted and reloaded its configuration, perhaps as a result of the upgrade? According to @YHNdnzj, udev DB is up and available during restarts normally but maybe there's a 255 vs 256 version incompatibility between the two and udevd 256 needs to regenerate the udev DB, hence why the devices dissappear temporarily.
```
Nov 22 22:10:35 twshared8602.05.cln3.facebook.com systemd[1]: systemd-udevd.service: Trying to enqueue job systemd-udevd.service/try-restart/fail
Nov 22 22:10:35 twshared8602.05.cln3.facebook.com systemd[1]: systemd-udevd.service: Installed new job systemd-udevd.service/restart as 4982
Nov 22 22:10:35 twshared8602.05.cln3.facebook.com systemd[1]: systemd-udevd.service: Enqueued job systemd-udevd.service/restart as 4982
Nov 22 22:10:35 twshared8602.05.cln3.facebook.com systemd-udevd[16701]: Using default interface naming scheme 'v255'.
Nov 22 22:10:35 twshared8602.05.cln3.facebook.com systemd[1]: systemd-udevd.service: Got notification message from PID 16701 (READY=1, STATUS=Processing with 120 children at max)
Nov 22 22:10:35 twshared8602.05.cln3.facebook.com systemd[1]: systemd-udevd.service: Changed start -> running
Nov 22 22:10:35 twshared8602.05.cln3.facebook.com systemd[1]: systemd-udevd.service: Job 4982 systemd-udevd.service/start finished, result=done
Nov 22 22:10:35 twshared8602.05.cln3.facebook.com systemd[1]: Started systemd-udevd.service - Rule-based Manager for Device Events and Files.
Nov 22 22:10:35 twshared8602.05.cln3.facebook.com systemd[1]: systemd-udevd.service: System call futex_waitv is not known, ignoring.
Nov 22 22:10:35 twshared8602.05.cln3.facebook.com systemd[1]: systemd-udevd.service: System call riscv_hwprobe is not known, ignoring.
Nov 22 22:10:35 twshared8602.05.cln3.facebook.com systemd[1]: /usr/lib/systemd/system/systemd-udevd.service:38: System call futex_waitv is not known, ignoring.
Nov 22 22:10:35 twshared8602.05.cln3.facebook.com systemd[1]: /usr/lib/systemd/system/systemd-udevd.service:38: System call riscv_hwprobe is not known, ignoring.
Nov 22 22:10:35 twshared8602.05.cln3.facebook.com systemd[1]: /usr/lib/systemd/system/systemd-udevd.service:38: System call fchmodat2 is not known, ignoring.
Nov 22 22:10:35 twshared8602.05.cln3.facebook.com systemd[1]: /usr/lib/systemd/system/systemd-udevd.service:38: System call set_mempolicy_home_node is not known, ignoring.
Nov 22 22:10:35 twshared8602.05.cln3.facebook.com (udev-worker)[18065]: Network interface NamePolicy= disabled on kernel command line.
Nov 22 22:10:35 twshared8602.05.cln3.facebook.com systemd[1]: systemd-udevd.service: Changed dead -> running
Nov 22 22:10:36 twshared8602.05.cln3.facebook.com systemd[1]: systemd-udevd.service: Got notification message from PID 16701 (RELOADING=1, MONOTONIC_USEC=328057670, STATUS=Reloading configuration...)
Nov 22 22:10:36 twshared8602.05.cln3.facebook.com systemd[1]: systemd-udevd.service: Changed running -> reload-notify
Nov 22 22:10:36 twshared8602.05.cln3.facebook.com systemd[1]: systemd-udevd.service: Installed new job systemd-udevd.service/nop as 5191
Nov 22 22:10:36 twshared8602.05.cln3.facebook.com systemd[1]: systemd-udevd.service: Job 5191 systemd-udevd.service/nop finished, result=done
Nov 22 22:10:36 twshared8602.05.cln3.facebook.com systemd[1]: systemd-udevd.service: Got notification message from PID 16701 (READY=1, STATUS=Processing with 120 children at max)
Nov 22 22:10:36 twshared8602.05.cln3.facebook.com systemd[1]: systemd-udevd.service: Changed reload-notify -> running
```

The devices then come back ~1 second later - example of device below:
```
Nov 22 22:10:28 twshared8602.05.cln3.facebook.com systemd[1]: dev-nvme0n1p2.device: Changed dead -> plugged
Nov 22 22:10:34 twshared8602.05.cln3.facebook.com systemd[1]: dev-nvme0n1p2.device: Changed dead -> plugged
Nov 22 22:10:35 twshared8602.05.cln3.facebook.com systemd[1]: dev-nvme0n1p2.device: Changed dead -> plugged
Nov 22 22:10:35 twshared8602.05.cln3.facebook.com systemd[1]: dev-nvme0n1p2.device: Changed plugged -> dead
Nov 22 22:10:36 twshared8602.05.cln3.facebook.com systemd[1]: dev-nvme0n1p2.device: Changed dead -> plugged
Nov 22 22:10:41 twshared8602.05.cln3.facebook.com systemd[1]: dev-nvme0n1p2.device: Changed dead -> plugged
Nov 22 22:10:41 twshared8602.05.cln3.facebook.com systemd[1]: dev-nvme0n1p2.device: Changed dead -> plugged
```

See mounts depending on this device:
```
[root@twshared8602 ~]# systemctl list-dependencies --reverse dev-nvme0n1p2.device
dev-nvme0n1p2.device
● ├─boot-vmlinux\x2d6.4.3\x2d0_fbk14_2601_gcd42476b84e9.mount
● ├─run-fs-control.mount
● ├─tmp.mount
● ├─usr-lib-modules-6.4.3\x2d0_fbk14_2601_gcd42476b84e9.mount
● ├─usr-src-kernels-6.4.3\x2d0_fbk14_2601_gcd42476b84e9.mount
● └─var-facebook-tupperware.mount
```

## TL;DR - Current theory

When PID1 reloads/reexecs, it calls manager_enumerate() to enumerate the devices in sysfs via device_enumerate(): https://github.com/systemd/systemd/blob/5b2926d9414f4333153ebe0bf169e1dd76129119/src/core/manager.c#L1853

However, if udev is reloading during the time, the manager may not find any devices registered in udev.

Then the manager will transition the devices from plugged -> dead, even if the mount is found in /proc/self/mountinfo: https://github.com/systemd/systemd/blob/5b2926d9414f4333153ebe0bf169e1dd76129119/src/core/device.c#L195

Since many of our mount units have implicit dependency on these devices (see systemd.mount man page), the mount units are stopped. And because the mount units are stopped, all units associated with the default target (e.g. chef-firstboot-os) are stopped. Once udev comes back up, the devices come back and all units are restarted. But since chef-firstboot-os is a oneshot service, its not restarted.

## Possible solutions

1. Don't do the restart of systemd-udevd and manager reload/reexec at the same time. Basically, we can take restarting udevd out of the RPM spec and restart it in Chef after the RPM upgrade is complete. While it doesn't fix the issue, it reduces the likelihood of a race here. This also may be good as its pretty rare to do systemd major version live upgrades like Meta does.
2. If the device is found in /proc/self/mountinfo but not udev, move it into DEVICE_TENTATIVE state, regardless of if it was previously seen by udev or not. Udev will eventually load and correct it. I put up a PR for this since its straightforward (but looking for feedback in case it has side effects): https://github.com/systemd/systemd/pull/35330
3. If many devices transition to dead on manager reload/reexec, udev may have been down. So ping udev waiting for it to be up, then enumerate the devices again. However, this adds a manager -> udevd RPC dependency which we may not want for distros that don't use udevd.

## Open questions I need to answer
1. Why do devices not show up in sysfs while udev is reloading during 255 -> 256 upgrade? Need to add debug logs to get this info.
2. Why does PID1 restart all targets and units associated with default target (e.g. chef-firstboot-os) instead of just the mounts? We don't use BindTo= or StopPropagatedFrom=
3. Why does this only happen on 255 -> 256 upgrade and not 255 -> 255? We did several of the latter upgrades with no issues.

### Unexpected behaviour you saw

_No response_

### Steps to reproduce the problem

_No response_

### Additional program output to the terminal or log subsystem illustrating the issue

_No response_

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #35329.

### CODE REVIEW

**Comment by yuwata** [comment]
If you want to take more stable way, then maybe better to revert 405be62f05d76f1845f347737b5972158c79dd3e entirely now. And later carefully add them one by one after v257.

**Review by YHNdnzj — APPROVED** [review]

**Review by ryantimwilson — APPROVED** [review]
Thank you for fixing and a reproducer!

File: test/units/TEST-17-UDEV.device_is_processing.sh [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+for _ in {1..3}; do
+    systemctl daemon-reload
+    systemctl is-active testsleep.service
+done
+
+# Check if the reexec and reload have finished during processing the event.
+grep -F 'ID_PROCESSING=1' "/run/udev/data/n${IFINDEX}"
+
+# cleanup
```
Comment (ryantimwilson) [inline test/units/TEST-17-UDEV.device_is_processing.sh]: Nit: Should we `trap` and add cleanup into trap in case the test fails in the middle? Something like https://github.com/systemd/systemd/blob/4f3df8c1bbdfb708a4d89144d20863f7fa51b121/test/units/TEST-07-PID1.private-pids.sh#L17-L35

But this is non-urgent and we need to get the PR in ASAP for 257-rc so feel free to ignore

File: src/core/device.c [hunk 2]
```diff
@@ -1048,9 +1048,6 @@ static void device_enumerate(Manager *m) {
                 _cleanup_set_free_ Set *ready_units = NULL, *not_ready_units = NULL;
                 Device *d;
 
-                if (device_is_processed(dev) <= 0)
-                        continue;
```
Comment (poettering) [inline src/core/device.c:1052]: so lemme get this right, ID_PROCESSING= is set when again?

after the main rules block completed a device, but before RUN= is completed. did i get this right? (we probably should add a comment somewher clarifying that).

So that would mean that a device with ID_PROCESSING=1 should probably *not* be considered present yet when a device first shows up, but *should* be considered present if it already was seen before. Right?

but that would suggest that we probably should consider the deseralized state here, shouldn't we?

i.e. if the deserialized state is active, we should not care about the ID_PROCESSING state of the device, but if the deserialized state is unknown or inactive then we should care about it, and skip the device as before, so that it will not be marked active here, until it actually is complete.

Or did i get this wrong? does this make any sense?

Comment (yuwata) [reply src/core/device.c:1052]: Everything right, I think. But that is slightly large in post rc2, so here the change is simply reverted.
I'd like to implement exactly what you commented after v257.

Comment (yuwata) [reply src/core/device.c:1052]: -> #35351
